src/bindings/python/flux/cli/plugin.py

```diff
@@ -0,0 +1,187 @@
+            values (:obj:`Dict`): Dictionary of KEY[=VALUE] pairs returned
+                by parsing the arguments provided to the -P/--plugin
+                option. KEY is None if plugin was not invoked. VALUE is the
+                empty string if the plugin was invoked without VALUE.
+        """
+        pass
+
+    def validate(self, jobspec):
+        """Allow a plugin to validate jobspec
+
+        This callback may be used by the cli itself or a job validator
+        to validate the final jobspec.
+
+        On an invalid jobspec, this callback should raise ValueError
+        with a useful error message.
+
+        Args:
+            jobspec (:obj:`flux.job.Jobspec`): jobspec object to validate.
+        """
+        pass
+
+
+class CLIPluginRegistry:
+    """Flux CLI plugin registry helper class"""
+
+    def __init__(self):
+        self.plugins = []
+        etc = conf_builtin_get("confdir")
+        if etc is None:
+            raise ValueError("failed to get builtin confdir")
+        if getenv("FLUX_CLI_PLUGINPATH"):
+            self.plugindir = getenv("FLUX_CLI_PLUGINPATH")
+        else:
+            self.plugindir = f"{etc}/cli/plugins"
+
+    def _add_plugins(self, module, program):
+        entries = [
+            getattr(module, attr) for attr in dir(module) if not attr.startswith("_")
+        ]
+        ## TODO: as we get each entry, make sure it is NOT speciifically the CLIPlugin base class
+        for entry in entries:
+            # only process entries that are an instance of type (i.e. a class)
+            # and are a subclass of CLIPlugin:
+            if isinstance(entry, type) and issubclass(entry, CLIPlugin):
+                self.plugins.append(entry(program))
+
+    def load_plugins(self, program):
+        """Load all cli plugins from the standard path"""
+        for path in glob.glob(f"{self.plugindir}/*.py"):
+            self._add_plugins(import_path(path), program)
+        return self
+
+    def add_plugin_option(self, parser):
+        """Add the -P option and list of available keys"""
+        parser.add_argument(
+            "-P",
+            "--plugin",
+            type=CLIPluginValue,
+            action="append",
+            help=f"{self.get_plugin_help_messages()}",
+        )
+
+    def get_plugin_options(self):
+        """Return all plugin option keys from self.plugins"""
+        opts = []
+        for plugin in self.plugins[1:]:
+            opts.append(str(plugin.get_plugin_name()))
+        return opts
+
+    def get_plugin_help_messages(self):
+        """Return a string that has all self.plugin help messages"""
+        ## TODO: Utilize the HelpFormatter class in argparse to make
+        ##  these messages better
+        help = ""
+        for plugin in self.plugins[1:]:
+            help += str(plugin.get_plugin_name())
+            help += str(plugin.get_help_message())
+            help += "\n"
+        return help
```

Comment:
This should be replaced with a function to pretty print the option+usage table as noted above.